sdk/python/kfp/dsl/_component_bridge.py

```diff
@@ -549,6 +556,49 @@ def _resolve_ir_placeholders_v2(
             'Input argument supports only the following types: PipelineParam'
             ', str, int, float. Got: "{}".'.format(argument_value))
 
+    # Make sure we're not passing parameter to artifact, or vice versa.
+    if isinstance(argument_value, _pipeline_param.PipelineParam):
+
+      # Pipeline function argument defaults to 'String' type if type is unknown.
+      if reference_type is None and argument_value.op_name is None:
+        reference_type = 'String'
+
+      # Loop arguments defaults to 'String' type if type is unknown.
+      if reference_type is None and isinstance(
+          argument_value,
+          (_for_loop.LoopArguments, _for_loop.LoopArgumentVariable)):
+        reference_type = 'String'
```

Comment:
Interesting that the tests pass (now).
Generally, the loop variable can be passed to inputs that are not strings (e.g. to number inputs).